File: clients/java/open-feature-provider/src/main/java/io/superposition/openfeature/SuperpositionProviderOptions.java
package io.superposition.openfeature;

import io.superposition.openfeature.options.EvaluationCacheOptions;
import io.superposition.openfeature.options.RefreshStrategy;
import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import org.jetbrains.annotations.Nullable;

@Builder
@Data
public class SuperpositionProviderOptions {
    @NonNull
    String endpoint;
    @NonNull
    String workspaceId;
    @NonNull
    String orgId;
    @NonNull
    String token;
    @NonNull
    RefreshStrategy refreshStrategy;
    @Nullable EvaluationCacheOptions evaluationCacheOptions;
    @Nullable SuperpositionConfig fallbackConfig;
    @Nullable ExperimentationOptions experimentationOptions;

    @Builder
    public static class ExperimentationOptions {
        @NonNull
        RefreshStrategy refreshStrategy;
        @Nullable
        EvaluationCacheOptions evaluationCacheOptions;
    }
}

File: clients/haskell/open-feature-provider/lib/Data/OpenFeature/SuperpositionProviderOptions.hs
module Data.OpenFeature.SuperpositionProviderOptions where

import Control.Monad.Logger (LogLevel (..))
import Data.Aeson (ToJSON)
import Data.Int (Int64)
import Data.Text (Text)
import GHC.Generics (Generic)
import Network.URI qualified as Net

data RefreshOptions = OnDemand Int64 | Poll Int
  deriving (Generic, Read, Show, ToJSON, Eq)

-- Orphans: LogLevel and URI come from libraries that do not provide these, and the options record
-- cannot derive Show/Eq/ToJSON without them.
deriving instance Generic LogLevel

deriving instance ToJSON LogLevel

deriving instance ToJSON Net.URIAuth

deriving instance ToJSON Net.URI

data SuperpositionProviderOptions = SuperpositionProviderOptions
  { orgId :: Text,
    workspaceId :: Text,
    endpoint :: Net.URI,
    token :: Text,
    -- TODO
    fallbackConfig :: (),
    refreshOptions :: RefreshOptions,
    experimentationRefreshOptions :: Maybe RefreshOptions,
    -- | Only fetch config keys under these prefixes. Nothing fetches the whole workspace config.
    configPrefixes :: Maybe [Text],
    logLevel :: LogLevel
  }
  deriving (Generic, Show, ToJSON, Eq)

defaultProviderOptions :: SuperpositionProviderOptions
defaultProviderOptions =
  SuperpositionProviderOptions
    { orgId = "",
      workspaceId = "",
      endpoint = Net.URI "" Nothing "" "" "",
      token = "",
      fallbackConfig = (),
      refreshOptions = OnDemand 0,
      experimentationRefreshOptions = Nothing,
      configPrefixes = Nothing,
      logLevel = LevelError
    }
